### PR TITLE
Added tags for tasks

### PR DESCRIPTION
**Pull request by Greed54**

### CODE REVIEW

File: src/main/java/com/list/todo/repositories/TaggedTaskRepository.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+import com.list.todo.entity.Tag;
+import com.list.todo.entity.TaggedTask;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import java.util.List;
+
+public interface TaggedTaskRepository extends JpaRepository<TaggedTask, Long> {
+    List<TaggedTask> findByTag(Tag tag);
+    TaggedTask findByTaskIdAndTag(Long taskId, Tag tag);
```
Comment (izluben) [inline src/main/java/com/list/todo/repositories/TaggedTaskRepository.java]: Why it is not Optional?

File: src/main/java/com/list/todo/services/TaggedTaskService.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    public List<TaggedTask> getTaggedTaskByTag(Tag tag) {
+        return taggedTaskRepository.findByTag(tag);
+    }
+
+    public void deleteTaggedTask(Long taskId, Tag tag) {
+
+        TaggedTask taggedTask = taggedTaskRepository.findByTaskIdAndTag(taskId, tag);
+
+        taggedTaskRepository.delete(taggedTask);
```
Comment (izluben) [inline src/main/java/com/list/todo/services/TaggedTaskService.java:73]: Will get an exception here if taggedTask is null

File: src/main/java/com/list/todo/services/TaggedTaskService.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        return myTaggedTask;
+    }
+
+    public Set<Task> getTasksByTags(List<Long> tagsIds, Long currentUserId) {
+        Set<Task> tasksByTags = new HashSet<>();
+
+        tagsIds.forEach(tagId -> {
+            Optional<Tag> tag = tagService.getTagById(tagId);
+
+            if (tag.isPresent()) {
```
Comment (izluben) [inline src/main/java/com/list/todo/services/TaggedTaskService.java]: Please, get rid of these two **if{}** statements using map(), filter() and ifPresent() instead

File: src/main/java/com/list/todo/controllers/TodoListController.java [hunk 4]
```diff
@@ -29,9 +30,10 @@
 
     @GetMapping("/my")
     public ResponseEntity<Iterable<TodoList>> getMyTodoLists(@AuthenticationPrincipal UserPrincipal currentUser,
+                                                             @RequestBody List<Long> tagsId,
```
Comment (izluben) [inline src/main/java/com/list/todo/controllers/TodoListController.java]: Ids, not Id

File: src/main/java/com/list/todo/graphql/TodoListQuery.java [hunk 5]
```diff
@@ -29,7 +29,7 @@
 
 	public Iterable<TodoList> getMyTodoLists() {
 		UserPrincipal currentUser = userService.getCurrentUser();
-		return todoListService.getTodoListsByUser(currentUser.getUsername(), Pageable.unpaged());
+		return null;
```
Comment (izluben) [inline src/main/java/com/list/todo/graphql/TodoListQuery.java]: Is it intentional change?

File: src/main/java/com/list/todo/entity/Tag.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+@Data
+@EqualsAndHashCode(callSuper = true)
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+@Builder
+public class Tag extends BaseEntity{
+
+    @NotNull
+    private String nameTag;
```
Comment (izluben) [inline src/main/java/com/list/todo/entity/Tag.java]: **tagName**, or just **name**

File: src/main/java/com/list/todo/controllers/TagController.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                                                     @RequestParam("taskId") Long taskId,
+                                                     @AuthenticationPrincipal UserPrincipal currentUser) {
+        ResponseEntity<Void> responseEntity;
+        Optional<Tag> tag = tagService.getTagById(tagId);
+
+        if (tag.isPresent()) {
+            if (!tag.get().getOwnerId().equals(currentUser.getId())) {
+                responseEntity = new ResponseEntity<>(HttpStatus.FORBIDDEN);
+            } else {
+                taggedTaskService.deleteTaggedTask(taskId, tag.get());
```
Comment (izluben) [inline src/main/java/com/list/todo/controllers/TagController.java]: Logic of removal should be moved into Service layer. Controller must not know that removal of a tag is actually removal of the DB table entry.

File: src/main/java/com/list/todo/entity/TaggedTask.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+import javax.validation.constraints.NotNull;
+
+@Entity
+@Data
+@EqualsAndHashCode(callSuper = true)
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString
+@Builder
+public class TaggedTask extends BaseEntity {
```
Comment (izluben) [inline src/main/java/com/list/todo/entity/TaggedTask.java:17]: Ill-fitted naming. TaggedTask  class implies that there are two types of tasks: TaggedTask and just Task (presumably not tagged), whereas in reality we have only one type of task: **Task**, and TaggedTask is only mapping

File: src/main/java/com/list/todo/controllers/TagController.java (hunk 7 above)
Comment (izluben) [reply src/main/java/com/list/todo/controllers/TagController.java]: Same applies to addTagToTask()

Comment (Greed54) [reply src/main/java/com/list/todo/controllers/TagController.java]: У нас же везде такие проверки в контроллерах, какую логику выносить в сервисы?

Comment (izluben) [reply src/main/java/com/list/todo/controllers/TagController.java]: you have TaskService. You can add method .deleteTag() into it, and do
` taggedTaskService.deleteTaggedTask(taskId, tag.get());` inside it

Comment (izluben) [reply src/main/java/com/list/todo/controllers/TagController.java]: Also use English in public conversations, that's kind of a company policy, not just my whim)))

**Pull request closed without merging**